plutus-contract/src/Plutus/Contract/Effects.hs

```diff
@@ -156,6 +165,80 @@ instance Pretty UtxoAtAddress where
       utxos = vsep $ fmap prettyTxOutPair (Map.toList utxo)
     in vsep ["Utxo at" <+> pretty address <+> "=", indent 2 utxos]
 
+-- | Validity of a transaction that has been added to the ledger
+data TxValidity = TxValid | TxInvalid | UnknownValidity
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving anyclass (ToJSON, FromJSON)
+  deriving Pretty via (PrettyShow TxValidity)
+
+instance MeetSemiLattice TxValidity where
+  TxValid /\ TxValid     = TxValid
+  TxInvalid /\ TxInvalid = TxInvalid
+  _ /\ _                 = UnknownValidity
+
+{- Note [TxStatus state machine]
+
+The status of a transaction is described by the following state machine.
+
+Current state | Next state(s)
+-----------------------------------------------------
+Unknown       | OnChain
+OnChain       | OnChain, Unknown, Committed
+Committed     | -
+
+The initial state after submitting the transaction is Unknown.
+
+-}
+
+-- | How many blocks deep the tx is on the chain
+newtype Depth = Depth Int
+    deriving stock (Eq, Ord, Show, Generic)
+    deriving newtype (Num, Real, Enum, Integral, Pretty, ToJSON, FromJSON)
+
+instance MeetSemiLattice Depth where
+  Depth a /\ Depth b = Depth (max a b)
+
+-- | The status of a Cardano transaction
+data TxStatus =
+  Unknown -- ^ The transaction is not on the chain. That's all we can say.
+  | TentativelyConfirmed Depth TxValidity -- ^ The transaction is on the chain, n blocks deep. It can still be rolled back.
+  | Committed TxValidity -- ^ The transaction is on the chain. It cannot be rolled back anymore.
```

Comment:
Do we really need `TxValidity` here? In the `Committed` constructor, we're sure that the transaction is valid no?

Reply:
When a transaction fails in phase 2 validation, it is still recorded on the blockchain. That's why we have the `OnChainTx` type with valid/invalid transactions.